### ISSUE TITLE
Align goals for CCM for official upstream cloudstack kubernetes provider?

### ISSUE THREAD

**Issue by yadvr** [tsuru/custom-cloudstack-ccm#15]
Hi guys,

Thanks for starting the work. The upstream kubernetes project will eventually remove the CloudStack k8s provider soon, see https://github.com/kubernetes/kubernetes/pull/68199
In that case, we'll need to host/maintain something similar to your ccm, probably as part of the upstream project. Can you advise your goals for the `ccm` and if they can align such that this can be developed and maintained under an upstream official CloudStack project repo?

**Comment by yadvr** [comment]
Ping @cezarsa @andrewsmedina @andrestc any updates guys?

**Comment by yadvr** [comment]
cc-ing for reference @ustcweizhou @sebgoa @PaulAngus @nvazquez @DaanHoogland

**Comment by cezarsa** [comment]
Sorry for taking so long to address this.

tl;dr: I don't really think this project is a good replacement for the deprecated cloudstack CCM, at least not how it is right now.

First of all this project was created to handle some very specific issues with our (Globo.com) private Cloudstack cloud, and with each release it ended up becoming more specific to our reality yet.
Each of these changes either add more complexity that wouldn't exist in a common Cloudstack installation (if such a thing exists) or break compatibility altogether.

I'll try listing the changes we have in this CCM compared to the old cloudstack one.

1 - We have multiple Cloudstack API endpoints with multiple Project IDs, and we wanted a CCM being able to talk to all these multiple instances on the same kubernetes cluster. Because of this constraint we require a few labels to be present on every node identifying its API endpoint and also identifying which projectID it's in.
Related to this same issue, every LoadBalancer service must include annotations identifying which endpoint will be used and which projectID will be used.

https://github.com/tsuru/custom-cloudstack-ccm/blob/c3cb42cb7ef236a524f32ed7260eeafea5b7714a/cloudstack/cloudstack.go#L41-L42

2 - We have the concept of node pools on our infrastructure, a `pool` label in each node will identify which pool it belongs. Pools are completely segregated on a network level. Because of this we wanted LoadBalancers to only include nodes from a single pool as their destinations, this adds another big change to how load balancers are created because we now also require a pool annotation on each LoadBalancer service.

https://github.com/tsuru/custom-cloudstack-ccm/blob/c3cb42cb7ef236a524f32ed7260eeafea5b7714a/cloudstack/cloudstack.go#L38

3 - Cloudstack API is kind of a non-standardized mess, and it's our fault that we added to this mess by changing some API command names and arguments on our private installation. To support our non-standard version we allow custom command names for some operations used in LoadBalancer creation, including associate-ip and dissociate-ip.

https://github.com/tsuru/custom-cloudstack-ccm/blob/c3cb42cb7ef236a524f32ed7260eeafea5b7714a/cloudstack/cloudstack.go#L54-L56

4 - This is related to 3, but the way we create LBs is really messed-up right now, our custom implementation accepts a `additionalPorts` argument to load balancer rules and a single rule is created, instead of creating a new rule for each port as the original Cloudstack CCM did.

All that said, it would be possible to change this CCM to work in a similar fashion to the old Cloudstack CCM if no custom configuration is provided. However, due to our unique constraints the code would probably look overly complex.

It was not our intention for this to become the default Cloudstack CCM when we started this project and it shows on how we didn't care for breaking compatibility in the past, changing this now would be feasible but not without some heavy work and I'm afraid we do not have enough hands to work on this right now. If someone from the community would like to take this however I would love to provide some guidance.

**Comment by DaanHoogland** [comment]
@cezarsa when you say "some heavy work", do you have an idea of the amount? (lines of code/methods/blocks to change, days/weeks/.. anykind of measure)

**Comment by onitake** [comment]
SWISS TXT also prepared a CloudStack CCM: https://github.com/swisstxt/cloudstack-cloud-controller-manager
We intend to keep it as generic as possible and possibly move it back under the k8s umbrella at some point.

For the moment, it's basically the code base of the old CCM, with some modifications from us to make it work more reliably.
I'm happy to discuss any possibilities for a common project; even with custom integrations, it may be useful to work on the CCM together.

The old cloud provider will be removed from the k8s tree soon, so I think it's important to have a good replacement: https://github.com/kubernetes/kubernetes/pull/72178
